### ISSUE TITLE
Linting unnecessary `if`/`else` in case of an early `return()`/`stop()`

### ISSUE THREAD

**Issue by IndrajeetPatil** [r-lib/lintr#2208]
In case of an early return or stop, the `else` clause is unnecessary and nested code can be avoided. 
Typically, this happens when the first `if` acts like a guard clause.

I am wondering if `redundant_ifelse_linter()` would be a good place to cover this case.

## Examples

Examples are with `return()`, but it is easy to see how the same would hold for `stop()`.

``` r
foo1 <- function(x) {
  if (x > 2L) {
    return("greater than 2")
  } else { # should produce lint
    # all code in this branch can be moved out of the `else()` clause
    return("not greater than 2")
  }
}

foo2 <- function(x) {
  if (x > 2L) {
    return("greater than 2")
  }
  
  return("not greater than 2")
}

identical(foo1(3L), foo2(3L))
#> [1] TRUE
identical(foo1(1L), foo2(1L))
#> [1] TRUE
```

<sup>Created on 2023-09-21 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by MichaelChirico** [comment]
this is really a style choice, so calling the return "unnecessary"/"redundant" is harsh IMO.

eliminating nesting like this is a good tool, but doing so every possible time often impedes readability. this should be a new linter.

**Comment by MichaelChirico** [comment]
BTW, we have `UnnecessaryNestingLinter()`, it works in cases where _one branch is not a stop_ (and similar). That means the branching can't be justified as increasing readability by parallelism -- "the branches are unequal" so the non-`stop()` branch can be unnested, e.g.

```r
if (x < 0) {
  stop("Must be positive!")
} else {
  ...
}
```

Lints because the `else` branch doesn't have a `stop()` or a `return()`. It could be unnested:

```r
if (x < 0) {
  stop("Must be positive!")
}
...
```

**Comment by MichaelChirico** [comment]
I don't think there's a static way to improve code here further than what's already available in `unnecessary_nesting_linter()`.

The only thing we could do is offer a 'strict' mode that requires one branch to be unnested when both have an 'exit call', but I think that's too large a source of false positives. It's a good idea when there's a "primary" branch that obviously should get less nesting, but very often the branches are "coequal" in which case unnesting harms readability & such an unnesting will be arbitrary churn.
